### ISSUE TITLE
Potential bug "FullyQualifiedNameProvider is a dependency, but not set" while using FullRepoManager

### ISSUE THREAD

**Issue by cheog** [Instagram/LibCST#689]
TLDR; I'm using `FullRepoManager.get_metadata_wrapper_for_path` to get a wrapper that I call `visit` on but I'm getting the error `FullyQualifiedNameProvider is a dependency, but not set`

**Goal and Design of the Code**
I am trying to find every all the usages of packages internal to my company within an arbitrary function. The code is below but the brief rundown is as follows: I have a FindUses visitor that looks at every name and attribute, gets the fully qualified name (the code uses relative imports) and checks to see if that name starts with the internal package name. I have another visitor called FindFuncDefs that looks for function definitions that have particular names. When it finds a function with one of the names of interest it visits that function with FindUses.

Im using the `FullRepoManager.get_metadata_wrapper_for_path` to get a wrapper and I call `visit` on that wrapper. I've used this method successfully with FullyQualifiedNameProvider and I was using the exact same arguments and everything (I used the up arrow in ipython so I know everything was exactly the same). The big difference is that I wasn't calling visit within another visit.

**Error**
The error I'm getting is getting triggered on line 1 of `FindUses.add_ref` because of the `self.get_metadata` call.

**Code**
I can't provide the code this is acting on because it's proprietary. I understand if there's nothing that can be done without that code but I figured someone here might know something more than me at least. 
```
KeyError: 'FullyQualifiedNameProvider is a dependency, but not set; did you forget a MetadataWrapper?'
``` 
It's triggering on 

```python
from libcst import CSTVisitor
from libcst.metadata import FullRepoManager, FullyQualifiedNameProvider
from libcst.helpers import get_full_name_for_node
from pathlib import Path

class FindUses(CSTVisitor):
    METADATA_DEPENDENCIES = (FullyQualifiedNameProvider,)
    def __init__(self):
        super().__init__()
        self.referenced: dict[str, set[str]] = {}

    def add_ref(self, node):
        qual_names = self.get_metadata(FullyQualifiedNameProvider, node)
        if len(qual_names) == 0:
            return
        qual_name = list(qual_names)[0]

        if not qual_name.name.startswith("<internal module name>"):
            return

        name = get_full_name_for_node(node)
        assert name is not None
        module = qual_name.name.removesuffix(name).rstrip(".")
        self.referenced[module] = self.referenced.setdefault(module, set()) | {name}

    def visit_Name(self, node):
        self.add_ref(node)
        return False

    def visit_Attribute(self, node):
        self.add_ref(node)
        return False


class FindFuncDefs(CSTVisitor):
    METADATA_DEPENDENCIES = (FullyQualifiedNameProvider,)
    def __init__(self, visitor: CSTVisitor, functions: set[str]):
        super().__init__()
        self.visitor = visitor
        self.func_names = functions

    def visit_FunctionDef(self, node):
        if get_full_name_for_node(node.name) not in self.func_names:
            return True
        node.body.visit(self.visitor)
        return False

all_paths = [str(path.relative_to(root)) for path in root.rglob("*.py")]
mgr = FullRepoManager(root, all_paths, {FullyQualifiedNameProvider})
wrapper = mgr.get_metadata_wrapper_for_path("certain/path/under/root/__init__.py")
use_finder = FindUses()
func_finder = FindFuncDefs(use_finder, "<function name>")
wrapper.visit(func_finder)
```

**Comment by zsol** [comment]
Yeah this isn't going to work, because metadata isn't available for your inner visitor (`FindUses`). `wrapper.visit` makes sure that `func_finder`'s metadata dependencies are resolved and available (see [here](https://github.com/Instagram/LibCST/blob/c44ff0500b52dd78716c36c7d1efd3016e20005f/libcst/metadata/wrapper.py#L199-L205), but nothing makes sure the same for `use_finder`. Now you can try and make that happen yourself, but a much better idea is to refactor your two visitors into one.

If I were you I'd merge `FindFuncDefs` and `FindUses` into one, and I'd also look at [matcher decorators](https://libcst.readthedocs.io/en/latest/matchers_tutorial.html#Matcher-Decorators). Something like

```
class FindUses(m.MatcherDecoratableVisitor):
    ...
   
    @m.call_if_inside(
        m.FunctionDef(name=m.MatchIfTrue(lambda n: get_full_name_for_node(n) in func_names))
    )
    @m.visit(m.Name | m.Attribute)
    def add_ref(self, node: cst.Name | cst.Attribute) -> bool:
        ...
```
